### DISCUSSION TITLE
Type of defaultValue expected by useController cannot be defined in reusable input

### DISCUSSION THREAD

**Discussion by slugalisk** [react-hook-form/react-hook-form#6443] [Q&A]
### Version Number

7.14.2

### Codesandbox/Expo snack

https://codesandbox.io/s/charming-northcutt-zyv9p?file=/src/App.tsx

### Steps to reproduce

1. Go to [https://codesandbox.io/s/charming-northcutt-zyv9p?file=/src/App.tsx](https://codesandbox.io/s/charming-northcutt-zyv9p?file=/src/App.tsx)
2. Scroll line 26
3. See the typescript error.

### Expected behaviour

The type (string) of the default value should be assignable to `defaultValue` because `name` can only be the key of a string type field value from type `T`. This broke in a recent update but it may only ever have worked because the typing used to be more lax.

It's not really relevant but from the input implementer's perspective this functions correctly - see lines 53 and 54 where the valid and invalid names are flagged correctly.

I looked through 50+ implementations of this library on GitHub and I didn't see anyone who managed to write a reusable type safe input. Everyone resorts to typing `control` as `any` or `Control<any>` or using `UseControlProps<any>` so there's clearly a fair bit of confusion around how this should be written. Assuming it's possible it would be helpful to have an example.

### Relevant log output

```shell
(property) defaultValue?: UnpackNestedValue<PathValue<T, Path<T>>>
Type 'string' is not assignable to type 'UnpackNestedValue<PathValue<T, Path<T>>>'.ts(2322)
controller.d.ts(21, 5): The expected type comes from property 'defaultValue' which is declared here on type 'UseControllerProps<T, Path<T>>'
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
I suppose you are comparing with v6 against v7, that's indeed is the case and I guess it's the trade-off we made to take on v7 for strict typing with input name with string template literal. 

I can see your type error caused due to missing typing with "name"  "bar"

```
interface FooFormData {
  foo: string;
}
```

at the moment, it's impossible to do a partial type check. @kotarella1110 if you want to include your thought as well. thanks

**Reply by slugalisk** [reply]
got it. thanks for your feedback.

**Comment by akomm** [comment]
API Docs:
https://react-hook-form.com/api/usecontroller
State, quote:
`[...] It's useful for creating reusable Controlled input.`

The TS example there only works, because it is *not reusable*. It depends on some specific Form input type. While it was not hard to figure out the issue and reason, its really contra-productive to make claims and providing examples which do the opposite. Makes people think they've missed something and waste time looking around...

@slugalisk
Example how you could do it without "any" including completion for "name" property, when using it later on (given you provided Input type to useForm):

```tsx
import {TextField, TextFieldProps} from "@mui/material"
import {
  useController,
  UseControllerProps,
  UseControllerReturn
} from "react-hook-form"

type Props<FormValues> = Omit<
  TextFieldProps,
  keyof UseControllerProps | keyof UseControllerReturn["field"] | "inputRef"
> &
  UseControllerProps<FormValues>

export function FormTextField<FormValues>({
  name,
  rules,
  control,
  defaultValue,
  shouldUnregister,
  ...inputProps
}: Props<FormValues>) {
  const {
    field: {ref, ...formProps},
    fieldState: {error}
  } = useController({name, rules, control, defaultValue, shouldUnregister})
  return (
    <div className="flex flex-col space-y-2">
      <TextField {...inputProps} {...formProps} inputRef={ref} />
      {error && (
        <div className="rounded border border-red-500 text-red-500 text-sm px-4 py-0.5">
          {error.message}
        </div>
      )}
    </div>
  )
}

```

I actually try to avoid the non-arrow function syntax and also adding generics to component's props. There might be issues and AFAIK there was time when it did not even work. Just to say, in TS 4.4.2 it works for me...

A minimal solution based on example:

```tsx
import { useForm, useController, UseControllerProps } from "react-hook-form";

type FormValues = {
  FirstName: string;
  aNumber: number
};

function Input<T>(props: UseControllerProps<T>) {
  const { field, fieldState } = useController(props);

  return (
    <div>
      <input {...field} placeholder={props.name} />
      <p>{fieldState.isTouched && "Touched"}</p>
      <p>{fieldState.isDirty && "Dirty"}</p>
      <p>{fieldState.invalid ? "invalid" : "valid"}</p>
    </div>
  );
}

export default function App() {
  const { handleSubmit, control } = useForm<FormValues>({
    defaultValues: {
      FirstName: ""
    },
    mode: "onChange"
  });
  const onSubmit = (data: FormValues) => console.log(data);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Input control={control} name="FirstName" rules={{ required: true }} />
      <Input control={control} name="aNumber" defaultValue={false} rules={{ required: true }} />
      <input type="submit" />
    </form>
  );
}
```

Try change type of aNumber to `boolean` and see the `defaultvalue` property

**Reply by eightyfive** [reply]
> [...] It's useful for creating reusable Controlled input.

Very misleading indeed. Quite the opposite.

**Comment by jeromeSH26** [comment]
Hi,
wondering what is the way to type `Controller` when using it as generic. Have created a generic Select component in order to type the option's value and return type.
See below example. When passing the return type to the Controller, ESLint is complaining Has to force "any" in the Controller to remove typing error.

```
export interface IOptions<T> {
  value: T;
  displayedLabel: string;
}

type ISelectFieldProps<T extends Array<string | number> | string | number> =
  SelectProps<T> & {
    name: string;
    control: Control;
    options: Array<IOptions<T extends Array<string | number> ? UnBox<T> : T>>;
    optionsAsCheckBox?: boolean;
  };

const SelectField = <T extends Array<string | number> | string | number>({
  name,
  control,
  defaultValue,
  label,
  options,
  optionsAsCheckBox = false,
  ...rest
}: ISelectFieldProps<T>): JSX.Element => {
  return (
    <Controller<FieldValues, T> <---changing T to any fix the issue, but then value and defaultValue are not typeof T...
      name={name}
      control={control}
      defaultValue={defaultValue}
      render={({ field: { onChange, onBlur, value } }) => (
        <Box>
          <InputLabel id={name}>{label}</InputLabel>
          <Select
            label={label}
            onChange={onChange}
            value={value}
            onBlur={onBlur}
            {...rest}
          >
            {options.map((option) => (
              <MenuItem key={option.value} value={option.value}>
                {optionsAsCheckBox && (
                  <Checkbox checked={value.indexOf(option.value) > -1} />
                )}
                <ListItemText primary={option.displayedLabel} />{" "}
              </MenuItem>
            ))}
          </Select>
        </Box>
      )}
    />
  );
};

export default SelectField;
```
**example to use it** 

```
<SelectField<number>
        name="age"
        control={control}
        defaultValue={0}
        label="Age"
        options={[
          {
            value: 0,
            displayedLabel: "",
          },
          {
            value: 10,
            displayedLabel: "Dix",
          },
          {
            value: 20,
            displayedLabel: "Vingt",
          },
          {
            value: 30,
            displayedLabel: "Trente",
          },
        ]}
      />
  
        <SelectField<Array<string>>
        name="ton"
        multiple
        control={control}
        input={<OutlinedInput label="Tag" />}
        renderValue={(selected) => selected.join(", ")}
        defaultValue={[]}
        label="Ton"
        options={[
          {
            value: "A",
            displayedLabel: "AAAA",
          },
          {
            value: "B",
            displayedLabel: "BBBB",
          },
          {
            value: "C",
            displayedLabel: "CCCCC",
          },
        ]}
      />

```

![image](https://user-images.githubusercontent.com/51024915/140518495-5ffb1cc7-867c-41c0-b3dc-f415a04edaa5.png)
